### PR TITLE
3/105 place items in map

### PR DESCRIPTION
**Pull request by mitsudz**
Adds inventory and item UI

### CODE REVIEW

**Review by Jonoym — CHANGES_REQUESTED** [review]
Only add students in your studio as reviewers

**Review by PHONGNGYUEN1 — APPROVED** [review]
Branch is good to merge now

**Review by BaileyS03 — CHANGES_REQUESTED** [review]
1. Is it possible to provide more testing on the files you've committed? (PlayerInventoryDisplay, KeyboardPlayerInput, ItemProximity, itemFactory) etc. 
Potions is has some coverage though.
2. In some files //TODOs have been left i.e., in the PlayerInventoryDisplay AbstractItem. Please ensure that these are removed
3. Ensure methods are protected where possible, I.e. Public AbstractPotion should be protected.

File: source/core/src/main/com/csse3200/game/components/tasks/ItemProximityTask.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
+package com.csse3200.game.components.tasks;
```
Comment (BaileyS03) [inline source/core/src/main/com/csse3200/game/components/tasks/ItemProximityTask.java:1]: The tests are better, I would still like to see ItemProximityTask tests if possible, please (I made some ones like this in the team 4 branch if you needed example tests for tasks)

File: source/core/src/main/com/csse3200/game/inventory/Inventory.java [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+public class Inventory implements InventoryInterface {
+    private final int capacity; // The maximum number of items the inventory can hold.
+    private int freeSlots; // The current number of available slots in the inventory.
+    private int nextIndex = 0; // The index where the next item can be stored.
+    // Maps item codes to sets of inventory indices where they are stored.
+    private TreeMap<Integer, TreeSet<Integer>> codeToIndices;
+    // Maps item names to sets of inventory indices where they are stored.
+    private TreeMap<String, TreeSet<Integer>> nameToIndices;
+    // Array representing the inventory, holding items or null values.
+    private AbstractItem[] inventory; // Array of actual items & null values
```
Comment (BaileyS03) [inline source/core/src/main/com/csse3200/game/inventory/Inventory.java:27]: A field should not duplicate the name of its containing class

File: source/core/src/main/com/csse3200/game/inventory/items/ConsumableItem.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * Constructs a ConsumableItem with given initial starting quantity and stack limit.
+     * <p><b>Note - a consumable item cannot be created without a limit/quantity specified.</b></p>
+     *
+     * @param name the name of the item
+     * @param itemCode the item code
+     * @param limit the stack limit of the item
+     * @param quantity the initial quantity for this item
+     */
+    public ConsumableItem(String name, int itemCode, int limit, int quantity) {
```
Comment (BaileyS03) [inline source/core/src/main/com/csse3200/game/inventory/items/ConsumableItem.java:23]: Should this be protected?

File: source/core/src/main/com/csse3200/game/entities/factories/ItemFactory.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import com.csse3200.game.inventory.items.food.Foods;
+import com.csse3200.game.inventory.items.potions.healingpotion.HealingPotion;
+import com.csse3200.game.physics.components.ColliderComponent;
+import com.csse3200.game.physics.components.PhysicsComponent;
+import com.csse3200.game.rendering.TextureRenderComponent;
+
+import static com.csse3200.game.physics.PhysicsLayer.OBSTACLE;
+
+
+public class ItemFactory {
```
Comment (BaileyS03) [inline source/core/src/main/com/csse3200/game/entities/factories/ItemFactory.java:17]: Should this be private to hide implicit public one?

File: source/core/src/main/com/csse3200/game/entities/factories/ItemFactory.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
+package com.csse3200.game.entities.factories;
+
+import com.badlogic.gdx.physics.box2d.BodyDef;
+import com.csse3200.game.ai.tasks.AITaskComponent;
+import com.csse3200.game.components.tasks.*;
+import com.csse3200.game.entities.Entity;
+import com.csse3200.game.inventory.items.AbstractItem;
+import com.csse3200.game.inventory.items.food.Foods;
+import com.csse3200.game.inventory.items.potions.healingpotion.HealingPotion;
+import com.csse3200.game.physics.components.ColliderComponent;
```
Comment (BaileyS03) [inline source/core/src/main/com/csse3200/game/entities/factories/ItemFactory.java:10]: Do you use this import?

File: source/core/src/main/com/csse3200/game/entities/factories/ItemFactory.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import com.csse3200.game.physics.components.PhysicsComponent;
+import com.csse3200.game.rendering.TextureRenderComponent;
+
+import static com.csse3200.game.physics.PhysicsLayer.OBSTACLE;
```
Comment (BaileyS03) [inline source/core/src/main/com/csse3200/game/entities/factories/ItemFactory.java:14]: Do you use this import?

File: source/core/src/test/com/csse3200/game/inventory/items/potions/AbstractPotionTest.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import com.csse3200.game.extensions.GameExtension;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(GameExtension.class)
+public class AbstractPotionTest  {
```
Comment (BaileyS03) [inline source/core/src/test/com/csse3200/game/inventory/items/potions/AbstractPotionTest.java:14]: I think this public can go but thats not that important

File: source/core/src/test/com/csse3200/game/inventory/items/potions/AbstractPotionTest.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    private HealingPotion healingPotion;
+
+    @BeforeEach
+    public void setUp() {
```
Comment (BaileyS03) [inline source/core/src/test/com/csse3200/game/inventory/items/potions/AbstractPotionTest.java:18]: I think these publics can go too.

File: source/core/src/main/com/csse3200/game/inventory/items/AbstractItem.java [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    /**
+     * Constructs an AbstractItem with the specified name and item code. Defaults to single
+     * non-stackable item.
+     *
+     * @param name name of the item
+     * @param itemCode unique item code (up to name)
+     *
+     * <p><strong>Note - all concrete subclasses must provide a unique itemCode</strong></p>
+     */
+    public AbstractItem(String name, int itemCode) {
```
Comment (BaileyS03) [inline source/core/src/main/com/csse3200/game/inventory/items/AbstractItem.java:31]: should this be protected?

File: source/core/src/main/com/csse3200/game/inventory/items/AbstractItem.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+     * @param name name of the item
+     * @param itemCode unique item code (up to name)
+     * @param limit the maximum size of the item stack
+     * @param quantity the initial starting quantity of the item
+     *
+     * @throws IllegalArgumentException if limit or quantity is negative, or quantity > limit
+     *
+     * <p><strong>Note - all concrete subclasses must provide a unique itemCode</strong></p>
+     */
+    public AbstractItem(String name, int itemCode, int limit, int quantity) {
```
Comment (BaileyS03) [inline source/core/src/main/com/csse3200/game/inventory/items/AbstractItem.java:50]: Should this also be protected?

File: source/core/src/main/com/csse3200/game/inventory/items/AbstractItem.java [hunk 11]
```diff
@@ -0,0 +1,172 @@
+package com.csse3200.game.inventory.items;
+
+import com.badlogic.gdx.graphics.Texture;
```
Comment (BaileyS03) [inline source/core/src/main/com/csse3200/game/inventory/items/AbstractItem.java:3]: Do you use this import?

**Review by BaileyS03 — CHANGES_REQUESTED** [review]
Please see the comments I left in my code :) If you disagree can you please tell me why so I can understand too please!

**Pull request merged**